qsk-macros/src/lower.rs

```diff
@@ -0,0 +1,143 @@
+use proc_macro2::Span;
+use proc_macro_error::abort_call_site;
+use syn::{LitBool, LitStr, Ident, Path, PathSegment, PathArguments};
+use syn::punctuated::Punctuated;
+
+use qsk_types;
+
+pub struct LayerRefByName {
+    pub(crate) path: Path,
+    pub(crate) name: LitStr,
+}
+
+pub struct TapToggle {
+    pub(crate) name: Path,
+    pub(crate) layer_ref: LayerRefByName,
+    pub(crate) tap_key: Path,
+}
+
+pub enum ControlCode {
+    Key(Path),
+    TapToggle(TapToggle),
+    Exit(Path),
+}
+
+impl From<&qsk_types::ControlCode> for ControlCode {
+    fn from(cc: &qsk_types::ControlCode) -> ControlCode {
+        match cc {
+            qsk_types::ControlCode::KeyMap(kc) => {
+                ControlCode::Key(
+                    keycode_path(&kc.to_string())
+                )
+            },
+            qsk_types::ControlCode::TapToggle(layer_ref, kc) => {
+                ControlCode::TapToggle(TapToggle{
+                    name: control_code_path("TapToggle"),
+                    layer_ref: match layer_ref {
+                        qsk_types::LayerRef::ByName(name) => {
+                            LayerRefByName{
+                                path: path_from_vec_str(vec!["qsk_types", "LayerRef", "ByName"]),
+                                name: LitStr::new(name, Span::call_site()),
+                            }
+                        },
+                        qsk_types::LayerRef::ByIndex(_) => {
+                            abort_call_site!("referencing layers by index is unsupported")
```

Comment:
TODO: don't abort here, only abort in the procedural macro itself since `proc_macro_error::abort` can strictly only be used in the actual function that takes a token stream and outputs a token stream. calling it here reduces testability because the `proc_macro_error::abort` call panics wildly.

Instead we should propagate a `Result` all the way to the actual proc macro function where it would be fine to use `proc_macro_error::abort`.